src/react/client.tsx

```diff
@@ -404,7 +425,81 @@ async function browserMutex<T>(
   const lockManager = window?.navigator?.locks;
   return lockManager !== undefined
     ? await lockManager.request(key, callback)
-    : await callback();
+    : await manualMutex(key, callback);
+}
+
+function getMutexValue(key: string): {
+  currentlyRunning: Promise<void> | null;
+  waiting: Array<() => Promise<void>>;
+} {
+  if ((globalThis as any).__convexAuthMutexes === undefined) {
+    (globalThis as any).__convexAuthMutexes = {} as Record<
+      string,
+      {
+        currentlyRunning: Promise<void>;
+        waiting: Array<() => Promise<void>>;
+      }
+    >;
+  }
+  let mutex = (globalThis as any).__convexAuthMutexes[key];
+  if (mutex === undefined) {
+    (globalThis as any).__convexAuthMutexes[key] = {
+      currentlyRunning: null,
+      waiting: [],
+    };
+  }
+  mutex = (globalThis as any).__convexAuthMutexes[key];
+  return mutex;
+}
+
+function setMutexValue(
+  key: string,
+  value: {
+    currentlyRunning: Promise<void> | null;
+    waiting: Array<() => Promise<void>>;
+  },
+) {
+  (globalThis as any).__convexAuthMutexes[key] = value;
+}
+
+async function enqueueCallbackForMutex(
+  key: string,
+  callback: () => Promise<void>,
+) {
+  const mutex = getMutexValue(key);
+  if (mutex.currentlyRunning === null) {
+    setMutexValue(key, {
+      currentlyRunning: callback().finally(() => {
+        const nextCb = getMutexValue(key).waiting.shift();
+        setMutexValue(key, {
+          ...getMutexValue(key),
+          currentlyRunning:
+            nextCb === undefined ? null : enqueueCallbackForMutex(key, nextCb),
+        });
+      }),
+      waiting: [],
+    });
+  } else {
+    setMutexValue(key, {
+      ...mutex,
+      waiting: [...mutex.waiting, callback],
+    });
+  }
+}
+
+async function manualMutex<T>(
```

Comment:
Open to suggestions on approach here. Over DMs we discussed creating an indefinitely long chain of promises, which also works but feels a little gross to me (I'm not sure if I have a chain of 100 resolved promises if they get collapsed and garbage collected).

I could also probably store this state on the auth provider instead of a global and get the same locking. Not sure if we dislike using a global.

Reply:
global makes sense, matches the other semantics